doc/developer/design/20230903_avro_doc.md

```diff
@@ -0,0 +1,322 @@
+naming Avro schema nodes in the `kgen` load generator. Its syntax and semantics
+are described here.
+
+An Avro schema node can appear in the following contexts:
+
+  1. The top-level node of the schema,
+  2. One of the fields of a record,
+  3. One of the branches of a union,
+  4. The value type of an array, or
+  5. The value type of a map.
+
+We define the "unambiguous name" of a schema node recursively as follows:
+
+ * If the node is of a [named
+   type](https://avro.apache.org/docs/1.10.2/spec.html#names) (record, enum, or
+   fixed) then its unambiguous name is the same as its "fullname" as defined by
+   the Avro spec; e.g. just its name if it is not part of a namespace, or its
+   namespace, followed by `.`, followed by its name if it is.
+ * Otherwise, if the node is the top-level node of the schema, its unambiguous
+   name is the empty string (we will never have to deal with this, as sinks in
+   Materialize can't have a schema that is just a timestamp field and nothing
+   else. Things are always at least nested inside a record.)
+ * Otherwise, if the node is a record field, its unambiguous name is the
+   record's unambiguous name (i.e., its fullname), followed by `::`, followed by
+   the field name. Note that you don't need to account for multiple levels of
+   nesting here-- if record `my.namespace.a` contains record `my.namespace.b`
+   which contains record `my.namespace.c` which contains field `f1`, you can
+   just call that field `my.namespace.c::f1`, unambiguously.
+ * Otherwise, if the node is one of the variants of a union, its unambiguous
+   name is the union's unambiguous name, followed by a `.`, followed by the
+   node's index within a union. I think we should actually change this and use
+   the type name instead (e.g. `union_unambiguous_name.long` instead of
+   `union_unambiguous_name.4`). That would work because unions are not allowed
+   to contain multiple variants of the same kind of type unless they are named
+   types.
+ * Otherwise, if the node is the value type of an array or map, its unambiguous
+   name is the array or map's unambiguous name, followed by `[]`.
+
+### User-facing documentation
+
+The details in the [Design](#design) section above will need to be worked into
+the Avro sink documentation. In particular, we should be sure to precisely
+explain Avro schema node specifiers and the order in which Materialize searches
+for sources of documentation.
+
+### Implementation
+
+#### Parser
+
+The extensions to the parser are straightforward.
+
+#### Planner
+
+The planner will "freeze" any comments that have been promoted to documentation
+by rewriting the `CREATE SINK` statement to contain those comments as explicit
+`DOC ON` options. This ensures that the Avro schema generated for a sink is
+deterministic—i.e., that future changes to comments on the columns and types
+used by the sink will *not* result in Materialize generating a different schema
+on restart.
```

Comment:
From what I am understanding, using comments on columns and types as sources for fallback docs is the reason for the potential non-determinism. If we are anyways freezing the comments from `DOC ON` options, does it make sense to limit the scope to only support that?

Reply:
On second thoughts, it won't be saving on much effort. We'll just have to clarify that docs cannot be updated after sink is created (for the time being at least till an alter command is available).

Reply:
>On second thoughts, it won't be saving on much effort.

Yeah, my thoughts too!